plugins/pulp_rpm/plugins/distributors/yum/publish.py

```diff
@@ -700,25 +700,58 @@ def _get_pkglist_to_publish(self, erratum_unit):
             'packages': [],
 
         }
+        return proto_pkglist
+
+    @staticmethod
+    def _module_dict_as_nsvca(md):
+        return (md['name'], md['stream'], int(md['version']), md['context'], md['arch'])
+
+    def _get_pkglists_to_publish(self, erratum_unit):
+        """
+        Make a lists of packages which will be listed in the published erratum.
 
+        Only packages contained in the repository being published will be included in this
+        errata's pkglist.
+
+        This merges multiple Pulp pkglists into a single pkglist for publication, unless a
+        module happens. Then a fresh list is used for the module.
+
+        :param erratum_unit: the erratum unit to analyze
+        :param type: pulp_rpm.plugins.db.models.Errata
+        :return: list of pkglist to publish
+        :rtype: list of dicts
+        """
+        pkglist_index = 0
+        repo_id = self.get_repo().id
+        new_pkglist = self._new_pkglist(repo_id, pkglist_index)
+        ret = [new_pkglist]
         if not self.repo_unit_nevra:
             # if repo_unit_nevra is empty, this repo has no RPMs, which makes filtering easy:
             # no pkglist will ever contain packages, so short out and return the empty pkglist
-            return new_pkglist
-
+            return ret
         seen_packages = set()
-
         # find all pkglists for erratum despite the repo_id associated with pkglist
         pkglists = models.ErratumPkglist.objects(errata_id=erratum_unit.errata_id)
         for pkglist in pkglists:
             for collection in pkglist['collections']:
+                module = collection.get('module')
+                if module:
+                    # inside a modular collection
+                    if self._module_dict_as_nsvca(module) not in self.repo_module_nsvca:
+                        # not from the repo being published
+                        continue
+                    pkglist_index += 1
+                    new_pkglist = self._new_pkglist(repo_id, pkglist_index)
+                    ret.append(new_pkglist)
+                    new_pkglist['module'] = module
                 for package in collection.get('packages', []):
                     if package['filename'] not in seen_packages:
                         seen_packages.add(package['filename'])
                         if models.NEVRA._fromdict(package) in self.repo_unit_nevra:
                             new_pkglist['packages'].append(package)
-
-        return new_pkglist
+                # ret[0] is special; all non-modular collections are aggregated there
+                new_pkglist = ret[0]
```

Comment:
I'm not sure I see how all non-modular collections get aggregated in this collection, unless all the non-modular collections are present before any modular ones. Can you elaborate?

Reply:
imagine the list of collections started with a modular collection; the first item in the `ret` list is an empty list of packages; `pkglist_index` gets bumped to `1` and a new, modular package list is created in `new_pkglist` and appended to the `ret` list, which now contains an empty list as the 0th item and the modular collection list of packages as the second item. After that, the `new_pkglist` points to the 0th element of `ret` again and the outermost loop repeats.

The invariant could be stated as:
* at any iteration, the 0th `ret` item is a list of all the non-modular collection packages encountered, accumulated
* every modular collection has a custom list, following the 0th `ret` item
* the `pkglist_index` is never decreased

Btw the 0th `ret` item can happen to have `pkglist_index > 1` which I hope is OK. After the naming update, the index name will be`collection_index`

Reply:
actually I'm wrong here, the `pkglist_index` of the aggregated collection will always be 0, but might not be present in the erratum (I hope I've got it right this time ;)